cmd/webhook/admission/admission.go

```diff
@@ -0,0 +1,218 @@
+package admission
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/dataclients/kubernetes/definitions"
+	admissionsv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	MetricNamespace = "routegroup_admission"
+	metricSubsystem = "admitter"
+)
+
+var (
+	labels = []string{"admitter", "operation", "group", "version", "resource", "sub_resource"}
+
+	totalRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "requests",
+		Help:      "Total number of requests to this admitter",
+	}, []string{"admitter"})
+	invalidRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "invalid_requests",
+		Help:      "Total number of requests to this admitter that couldn't be parsed",
+	}, []string{"admitter"})
+	rejectedRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "rejected_admissions",
+		Help:      "Total number of requests rejected by this admitter",
+	}, labels)
+	approvedRequests = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "successful_admissions",
+		Help:      "Total number of requests successfully processed by this admitter",
+	}, labels)
+	admissionDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: MetricNamespace,
+		Subsystem: metricSubsystem,
+		Name:      "admission_duration",
+		Help:      "Duration of admission calls",
+		Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, .75, 1, 1.25, 1.5, 2, 2.5, 5, 10},
+	}, labels)
+)
+
+type Admitter interface {
+	Name() string
+	Admit(req *admissionsv1.AdmissionRequest) (*admissionsv1.AdmissionResponse, error)
+}
+
+type RouteGroupAdmitter struct {
+}
+
+func init() {
+	prometheus.MustRegister(totalRequests, invalidRequests, rejectedRequests, approvedRequests, admissionDuration)
+}
+
+func (r RouteGroupAdmitter) Name() string {
+	return "routegroup"
+}
+
+func (r RouteGroupAdmitter) Admit(req *admissionsv1.AdmissionRequest) (*admissionsv1.AdmissionResponse, error) {
+	rgItem := definitions.RouteGroupItem{}
+	err := json.Unmarshal(req.Object.Raw, &rgItem)
+	if err != nil {
+		emsg := fmt.Errorf("could not parse RouteGroup, %w", err)
```

Comment:
as I see, we never use emsg as an error object. We could just simply do:

```
emsg := fmt.Sprintf("could not parse RouteGroup, %v", err)
```

...and then use it as a string.